Add PropertyGrid to Design dialogs

bparks13 opened the pull request:
For all custom dialogs, there is now an added tab that contains a PropertyGrid to display all properties for a device. This is always the non-default tab, so that the priority is on the custom GUI, but it is always present should the user want to look at and change any properties that are not exposed via custom controls.

Any changes made in either the custom GUI or the PropertyGrid are operating on the same underlying object, which is the object attached to the PropertyGrid, so that all data is always synchronized.

**Neuropixels V1**
<img width="1002" height="691" alt="image" src="https://github.com/user-attachments/assets/de13dbeb-e11a-4e54-9545-c8b401317202" />

**Neuropixels V2**
<img width="1002" height="691" alt="image" src="https://github.com/user-attachments/assets/eec2e204-2758-4b9e-ad76-d432b916396f" />

**Rhs2116**
<img width="1002" height="691" alt="image" src="https://github.com/user-attachments/assets/134fa3e0-48a9-4256-a0c6-41c623459939" />

**Headstage64**
<img width="890" height="528" alt="image" src="https://github.com/user-attachments/assets/3e680ab8-46c4-4930-966f-aeff5ec6917c" />

<img width="890" height="528" alt="image" src="https://github.com/user-attachments/assets/fb45df48-e2ed-4095-bc0c-42fe72d450dd" />

Fixes #515

bparks13 committed:
Add PropertyGrid to stimulus sequence dialogs

- Remove constructor arguments from base class
- Rhs2116 is not refactored, as the controls are not 1:1 tied to an underlying object, but are dynamic controls depending on what is selected

bparks13 committed:
Add PropertyGrid to NeuropixelsV1 dialogs

bparks13 committed:
Add PropertyGrid to NeuropixelsV2 dialogs

jonnew reviewed (changes requested):
As we discussed during live code review, please move the ProbeGroup editor into the the specialized stimulus waveform property editor on the RHS stim gui so that the base class does not have to be informed about its existence.

bparks13 committed:
Filter for top level menu strips to avoid inconsistent menu items

bparks13 committed:
Add copy constructors to main library classes

bparks13 committed:
Address review comments

- Add categories to properties
- Allow the GUI panels to be resized so the property grid can be better visualized
- Fix the stimulus sequence options so that they resize properly
- Adjust NeuropixelsV2Dialog to show device properties at all times, and display probe configuration dialogs when selected from the property grid

bparks13 committed:
Add scaled channel configuration dialog

bparks13 committed:
Revert change to disable ellipsis

- NPX V2 probe configuration dialogs are not able to be opened from NPX V2e device dialogs

bparks13 committed:
Update NeuropixelsV2e dialog behavior

- Close the probe configuration dialog when any property other than ProbeConfiguration is selected
- Add text to panel giving instructions for the user on what to do to show the probe configuration dialog

bparks13 committed:
Add comment

bparks13 committed:
Add ShowInCustomDialogAttribute

- Hide certain properties when the property grid is displayed in a custom GUI, specifically hiding the DeviceName/DeviceAddress properties when a dialog comes from a ConfigureHeadstage* node to match the behavior of the Bonsai editor

bparks13 committed:
Allow probe configuration dialogs to be opened

bparks13 committed:
Hide properties tab for NeuropixelsV2e probe configurations when shown in the device dialog

bparks13 committed:
Update attribute name to DeviceTableProperty

bparks13 committed:
Rename file

jonnew commented:
When some property other than the probeconfiguration is selected, we get a helpful dialog in the area where the probe would be shown. I like this but I want a couple changes.

1. When I resize the form, the text is does not remain centered
    
   <img width="411" height="616" alt="image" src="https://github.com/user-attachments/assets/7bbf68c9-878c-4d4c-b87b-07980d15fe37" />

1. When I select a probe configuration, but I don't have a calibration file loaded, I would like the text to change to "Please provide a calibration file to configure your probe"
   
   <img width="1039" height="563" alt="image" src="https://github.com/user-attachments/assets/556caa5c-c7b7-4e5f-b426-6c3f79702d83" />

jonnew commented on OpenEphys.Onix1/ConfigureNeuropixelsV2e.cs:
```diff
@@ -36,6 +36,15 @@ public ConfigureNeuropixelsV2e(ConfigureNeuropixelsV2e configureNode)
+        /// Returns a deep copy of the current <see cref="ConfigureNeuropixelsV2e"/> instance.
+        /// </summary>
+        /// <returns><see cref="ConfigureNeuropixelsV2e"/> instance.</returns>
+        public IConfigureNeuropixelsV2 Clone()
```
Why are we adding this to the public interface instead of using the copy constructor like all the other configuration operations do? It looks like you are immediately interpreting the result as concrete `ConfigureNeuropixelsV2e` type anyway?

jonnew commented on OpenEphys.Onix1/ConfigureHeadstageNeuropixelsV2eBeta.cs:
```diff
@@ -37,6 +37,19 @@ public ConfigureHeadstageNeuropixelsV2eBeta()
+            Name = configureNode.Name;
+            Port = configureNode.Port;
+            PortControl = configureNode.PortControl;
+            NeuropixelsV2eBeta = configureNode.NeuropixelsV2eBeta.Clone() as ConfigureNeuropixelsV2eBeta ?? throw new InvalidOperationException($"Unable to copy {nameof(NeuropixelsV2eBeta)} property. Could not cast to the correct type.");
```
Something seems wrong here. The most superficial comment is why are you using `as` if you immediately throw an exception if the result is `null`? If you want to throw if the result of `configureNode.NeuropixelsV2eBeta.Clone()` is not castable to ConfigureNeuropixelsV2eBeta, then just use a cast

```
(ConfigureNeuropixelsV2eBeta)configureNode.NeuropixelsV2eBeta.Clone()
```

since this will throw in that case. The second issue is what's up with the `Clone()` method? I have a feeling this was not done at random but I feel like we need to explore if its required or if there is a better way to get the behavior you need.

jonnew reviewed (changes requested):
Lets have a live review of this one, I'm not getting the `clone()` stuff.

bparks13 committed:
Update Neuropixels dialogs

- Fix the device dialog label to remain centered when resizing
- Add label to probe configuration dialogs (V1/V2) with instructions on how to open the probe configuration

bparks13 committed:
Simplify copy constructor

jonnew reviewed (changes requested):
I think we should explore alternatives to this strategy. My big issue with this approach is that its polluting the main library for the sake of a feature we want in the GUIs. The only reason this might be a good idea is that the GUIs are pointing to a desirable feature in the public API that we have not implemented. But, I don't see that happening here. I see these changes as only in service of supporting the specific functionality of the GUI.

So, what's the alternative to get a deep copy for temporarily objects in the GUI? I think one that sounds crazy but actually is quite elegant is to use the Json serilize that you already have access to as a dependency. Something like:

``` csharp
using Newtonsoft.Json;

namespace OpenEphys.Onix1.Design
{
    internal static class DeepCopyHelper
    {
        public static T SerializeDeserialize<T>(T source)
        {
            var json = JsonConvert.SerializeObject(source);
            return JsonConvert.DeserializeObject<T>(json);
        }
    }
}
```
and 

``` csharp
 var configuration = DeepCopyHelper.SerializeDeserialize(configureHeadstage);
```

(using `NeuropixelsV1eHeadstageEditor` as an example) seems like it might fulfil your requirements without any modifications to the base library at all.

bparks13 committed:
Merge branch 'main' into issue-515
